framework/src/main/java/org/checkerframework/common/value/util/NumberUtils.java

```diff
@@ -32,24 +46,81 @@ public static List<? extends Number> castNumbers(
     }
     switch (typeKind) {
       case BYTE:
-        return CollectionsPlume.mapList(Number::byteValue, numbers);
+        if (isUnsigned) {
+          return CollectionsPlume.<Number, Short>mapList(
+              NumberUtils::byteValueUnsigned, (Iterable<Number>) numbers);
+        } else {
+          return CollectionsPlume.mapList(Number::byteValue, numbers);
+        }
       case CHAR:
         return CollectionsPlume.mapList(Number::intValue, numbers);
       case DOUBLE:
         return CollectionsPlume.mapList(Number::doubleValue, numbers);
       case FLOAT:
         return CollectionsPlume.mapList(Number::floatValue, numbers);
       case INT:
-        return CollectionsPlume.mapList(Number::intValue, numbers);
+        if (isUnsigned) {
+          return CollectionsPlume.<Number, Long>mapList(
+              NumberUtils::intValueUnsigned, (Iterable<Number>) numbers);
+        } else {
+          return CollectionsPlume.mapList(Number::intValue, numbers);
+        }
       case LONG:
         return CollectionsPlume.mapList(Number::longValue, numbers);
       case SHORT:
-        return CollectionsPlume.mapList(Number::shortValue, numbers);
+        if (isUnsigned) {
+          return CollectionsPlume.<Number, Integer>mapList(
+              NumberUtils::shortValueUnsigned, (Iterable<Number>) numbers);
+        } else {
+          return CollectionsPlume.mapList(Number::shortValue, numbers);
+        }
       default:
         throw new UnsupportedOperationException(typeKind + ": " + type);
     }
   }
 
+  /**
+   * Returns the given number, casted to unsigned byte.
+   *
+   * @param n a number
+   * @return the given number, casted to unsigned byte
+   */
+  private static Short byteValueUnsigned(Number n) {
+    short result = n.byteValue();
+    if (result > Byte.MAX_VALUE) {
```

Comment:
`result > Byte.MAX_VALUE` is never true.  (Same comment for the other *ValueUnsigned methods.